File: C-OS 4.0.8 alpha/tools/inject_storage.py
#!/usr/bin/env python3
"""Inject host files into a C-OS raw storage image.

Usage:
  tools/inject_storage.py --image build/storage.img \
      --add sample.mp3=/music/sample.mp3

The image format matches src/drivers/disk/storage.c. Both primary and backup
catalogs are updated and checksummed. Existing entries are replaced by a new
contiguous allocation; the old extent is intentionally left untouched because
this tool is for deterministic test images.
"""
import argparse
import os
import struct
import sys
import time
import zlib
from pathlib import PurePosixPath

SECTOR = 512
MAGIC = 0x43535350464F5354
VERSION = 4
DISK_SECTORS = 1048576
CATALOG_SECTOR = 0
CATALOG_SECTORS = 64
BACKUP_CATALOG_SECTOR = 65
DATA_START_SECTOR = 132
META_RING_SECTORS = 4096
MAX_FILES = 128
MAX_PATH = 128
HEADER_SIZE = 72
ENTRY_SIZE = 188
ENTRY_CHECKSUM_OFFSET = 180
CATALOG_CHECKSUM_OFFSET = 64
CATALOG_SIZE = HEADER_SIZE + MAX_FILES * ENTRY_SIZE


def get64(buf, off):
    return struct.unpack_from("<Q", buf, off)[0]


def put64(buf, off, value):
    struct.pack_into("<Q", buf, off, value)


def crc_skip(data, skip, length):
    crc = 0xFFFFFFFF
    crc = zlib.crc32(data[:skip], crc)
    crc = zlib.crc32(data[skip + length:], crc)
    return (~crc) & 0xFFFFFFFF


def normalize(path):
    path = path.replace("\\", "/")
    if not path.startswith("/"):
        path = "/" + path
    path = str(PurePosixPath(path))
    return "/" if path == "." else path


def entry(catalog, index):
    start = HEADER_SIZE + index * ENTRY_SIZE
    # A memoryview is required here: bytearray slicing returns a copy, which
    # would make injected entries disappear when the catalog is written.
    return memoryview(catalog)[start:start + ENTRY_SIZE]


def entry_path(e):
    return bytes(e[52:180]).split(b"\0", 1)[0].decode("utf-8", "replace")


def find_entry(catalog, path):
    for i in range(MAX_FILES):
        e = entry(catalog, i)
        if e[0] and entry_path(e) == path:
            return i
    return -1


def free_slot(catalog):
    for i in range(MAX_FILES):
        if entry(catalog, i)[0] == 0:
            return i
    return -1


def load_catalog(f):
    f.seek(CATALOG_SECTOR * SECTOR)
    c = bytearray(f.read(CATALOG_SIZE))
    if len(c) != CATALOG_SIZE:
        raise RuntimeError("catalog is truncated")
    header = [get64(c, off) for off in (0, 8, 16, 24, 32, 40)]
    expected = [MAGIC, VERSION, DISK_SECTORS, CATALOG_SECTORS, DATA_START_SECTOR, MAX_FILES]
    if header != expected:
        raise RuntimeError("storage catalog header mismatch: %r" % (header,))
    saved = get64(c, CATALOG_CHECKSUM_OFFSET)
    put64(c, CATALOG_CHECKSUM_OFFSET, 0)
    if saved != crc_skip(c, CATALOG_CHECKSUM_OFFSET, 8):
        raise RuntimeError("storage catalog checksum mismatch")
    put64(c, CATALOG_CHECKSUM_OFFSET, saved)
    return c


def ensure_dir(catalog, path, now):
    path = normalize(path)
    if path == "/" or find_entry(catalog, path) >= 0:
        return
    parent = normalize(str(PurePosixPath(path).parent))
    ensure_dir(catalog, parent, now)
    idx = free_slot(catalog)
    if idx < 0:
        raise RuntimeError("storage catalog has no free directory entry")
    e = entry(catalog, idx)
    e[:] = b"\0" * ENTRY_SIZE
    e[0], e[1] = 1, 1
    put64(e, 28, now)
    put64(e, 36, now)
    put64(e, 44, now)
    raw = path.encode("utf-8")[:MAX_PATH - 1]
    e[52:52 + len(raw)] = raw
    put64(e, ENTRY_CHECKSUM_OFFSET, crc_skip(e, ENTRY_CHECKSUM_OFFSET, 8))


def next_data_sector(catalog):
    result = DATA_START_SECTOR
    for i in range(MAX_FILES):
        e = entry(catalog, i)
        if not e[0] or e[1] != 0:
            continue
        result = max(result, get64(e, 4) + get64(e, 12))
    return result


def inject(image, specs):
    with open(image, "r+b") as f:
        if os.fstat(f.fileno()).st_size < DISK_SECTORS * SECTOR:
            raise RuntimeError("image is smaller than the C-OS storage format")
        catalog = load_catalog(f)
        now = int(time.time())
        next_sector = next_data_sector(catalog)
        added_bytes = 0
        for spec in specs:
            if "=" not in spec:
                raise RuntimeError("--add must be host-file=/path/in/C-OS: %s" % spec)
            host, dest = spec.split("=", 1)
            dest = normalize(dest)
            with open(host, "rb") as src:
                data = src.read()
            if not data:
                raise RuntimeError("refusing to inject an empty file: %s" % host)
            ensure_dir(catalog, normalize(str(PurePosixPath(dest).parent)), now)
            idx = find_entry(catalog, dest)
            if idx < 0:
                idx = free_slot(catalog)
            if idx < 0:
                raise RuntimeError("storage catalog has no free file entry")
            sectors = (len(data) + SECTOR - 1) // SECTOR
            if next_sector + sectors >= DISK_SECTORS - META_RING_SECTORS:
                raise RuntimeError("storage data area is full")
            padded = data + b"\0" * (sectors * SECTOR - len(data))
            f.seek(next_sector * SECTOR)
            f.write(padded)
            e = entry(catalog, idx)
            e[:] = b"\0" * ENTRY_SIZE
            e[0], e[1] = 1, 0
            put64(e, 4, next_sector)
            put64(e, 12, sectors)
            put64(e, 20, len(data))
            put64(e, 28, now)
            put64(e, 36, now)
            put64(e, 44, now)
            raw = dest.encode("utf-8")[:MAX_PATH - 1]
            e[52:52 + len(raw)] = raw
            put64(e, ENTRY_CHECKSUM_OFFSET, crc_skip(e, ENTRY_CHECKSUM_OFFSET, 8))
            print("injected %s -> %s (%d bytes, %d sectors)" %
                  (host, dest, len(data), sectors))
            next_sector += sectors
            added_bytes += len(data)
        used_files = sum(1 for i in range(MAX_FILES) if entry(catalog, i)[0])
        put64(catalog, 48, used_files)
        put64(catalog, 56, get64(catalog, 56) + added_bytes)
        put64(catalog, CATALOG_CHECKSUM_OFFSET, 0)
        put64(catalog, CATALOG_CHECKSUM_OFFSET, crc_skip(catalog, CATALOG_CHECKSUM_OFFSET, 8))
        f.seek(CATALOG_SECTOR * SECTOR)
        f.write(catalog)
        f.seek(BACKUP_CATALOG_SECTOR * SECTOR)
        f.write(catalog)
        verify = load_catalog(f)
        print("storage injection verified: %s (%d entries)" % (image, get64(verify, 48)))


def main():
    ap = argparse.ArgumentParser(description="inject files into a C-OS raw storage image")
    ap.add_argument("--image", default="build/storage.img")
    ap.add_argument("--add", action="append", required=True,
                    help="host-file=/path/in/C-OS; repeatable")
    args = ap.parse_args()
    try:
        inject(args.image, args.add)
    except (OSError, RuntimeError, ValueError) as exc:
        print("storage injection failed: %s" % exc, file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: C-OS 4.0.8 alpha/tools/inject_storage.go
package main

import (
    "encoding/binary"
    "errors"
    "flag"
    "fmt"
    "hash/crc32"
    "io"
    "os"
    "path/filepath"
    "strings"
    "time"
)

const (
    sectorSize = 512
    storageMagic uint64 = 0x43535350464f5354
    storageVersion uint64 = 4
    diskSectors uint64 = 1048576
    catalogSector uint64 = 0
    catalogSectors uint64 = 64
    backupCatalogSector uint64 = 65
    passwordSector uint64 = 129
    settingsSector uint64 = 130
    reservedSector uint64 = 131
    dataStartSector uint64 = 132
    metaRingSectors uint64 = 4096
    maxFiles = 128
    maxPath = 128
    headerSize = 72
    entrySize = 188
    catalogSize = headerSize + maxFiles*entrySize
    checksumOffset = 64
    entryChecksumOffset = 180
)

var crcTable = crc32.MakeTable(crc32.IEEE)

type addList []string
func (a *addList) String() string { return strings.Join(*a, ",") }
func (a *addList) Set(v string) error { *a = append(*a, v); return nil }

func u64(b []byte, off int) uint64 { return binary.LittleEndian.Uint64(b[off:off+8]) }
func put64(b []byte, off int, v uint64) { binary.LittleEndian.PutUint64(b[off:off+8], v) }
func checksumSkip(data []byte, skip, length int) uint64 {
    crc := uint32(0xffffffff)
    for i, x := range data {
        if i >= skip && i < skip+length { continue }
        crc = crc32.Update(crc, crcTable, []byte{x})
    }
    return uint64(^crc)
}
func pathBytes(dst []byte, path string) {
    for i := range dst { dst[i] = 0 }
    path = strings.ReplaceAll(path, "\\", "/")
    if !strings.HasPrefix(path, "/") { path = "/" + path }
    if len(path) >= len(dst) { path = path[:len(dst)-1] }
    copy(dst, path)
}
func entryPath(e []byte) string { return strings.TrimRight(string(e[52:180]), "\x00") }
func entryChecksum(e []byte) uint64 { return checksumSkip(e, entryChecksumOffset, 8) }
func catalogChecksum(c []byte) uint64 { return checksumSkip(c, checksumOffset, 8) }
func readAtFull(f *os.File, off int64, dst []byte) error {
    n, err := f.ReadAt(dst, off)
    if err != nil && !(err == io.EOF && n == len(dst)) { return err }
    if n != len(dst) { return io.ErrUnexpectedEOF }
    return nil
}
func writeAtFull(f *os.File, off int64, src []byte) error {
    n, err := f.WriteAt(src, off)
    if err != nil { return err }
    if n != len(src) { return io.ErrShortWrite }
    return nil
}
func fail(format string, args ...any) error { return fmt.Errorf(format, args...) }

func loadCatalog(f *os.File) ([]byte, error) {
    c := make([]byte, catalogSize)
    if err := readAtFull(f, int64(catalogSector*sectorSize), c); err != nil { return nil, err }
    if u64(c, 0) != storageMagic || u64(c, 8) != storageVersion ||
       u64(c, 16) != diskSectors || u64(c, 24) != catalogSectors ||
       u64(c, 32) != dataStartSector || u64(c, 40) != maxFiles {
        return nil, errors.New("storage catalog header mismatch")
    }
    saved := u64(c, checksumOffset); put64(c, checksumOffset, 0)
    if saved != catalogChecksum(c) { return nil, errors.New("storage catalog checksum mismatch") }
    put64(c, checksumOffset, saved)
    return c, nil
}
func findEntry(c []byte, path string) int {
    for i := 0; i < maxFiles; i++ {
        e := c[headerSize+i*entrySize:headerSize+(i+1)*entrySize]
        if e[0] != 0 && entryPath(e) == path { return i }
    }
    return -1
}
func usedEntryCount(c []byte) uint64 {
    var n uint64
    for i := 0; i < maxFiles; i++ { if c[headerSize+i*entrySize] != 0 { n++ } }
    return n
}
func ensureDir(c []byte, path string, now uint64) error {
    if path == "" || path == "/" { return nil }
    if findEntry(c, path) >= 0 { return nil }
    parent := filepath.ToSlash(filepath.Dir(path)); if parent == "." { parent = "/" }
    if err := ensureDir(c, parent, now); err != nil { return err }
    slot := -1
    for i := 0; i < maxFiles; i++ { if c[headerSize+i*entrySize] == 0 { slot = i; break } }
    if slot < 0 { return errors.New("storage catalog has no free entry for directory") }
    e := c[headerSize+slot*entrySize:headerSize+(slot+1)*entrySize]
    for i := range e { e[i] = 0 }
    e[0], e[1] = 1, 1
    put64(e, 28, now); put64(e, 36, now); put64(e, 44, now)
    pathBytes(e[52:180], path); put64(e, entryChecksumOffset, entryChecksum(e))
    return nil
}
func nextDataSector(c []byte) uint64 {
    next := dataStartSector
    for i := 0; i < maxFiles; i++ {
        e := c[headerSize+i*entrySize:headerSize+(i+1)*entrySize]
        if e[0] == 0 || e[1] != 0 { continue }
        end := u64(e, 4) + u64(e, 12)
        if end > next { next = end }
    }
    return next
}
func inject(image string, additions []string) error {
    f, err := os.OpenFile(image, os.O_RDWR, 0)
    if err != nil { return err }
    defer f.Close()
    info, err := f.Stat(); if err != nil { return err }
    if info.Size() < int64(diskSectors*sectorSize) { return fail("image too small: %d bytes", info.Size()) }
    c, err := loadCatalog(f); if err != nil { return err }
    now := uint64(time.Now().Unix())
    next := nextDataSector(c)
    var addedBytes uint64
    for _, spec := range additions {
        parts := strings.SplitN(spec, "=", 2)
        if len(parts) != 2 { return fail("--add expects host-file=/path/in/C-OS: %s", spec) }
        host, dest := parts[0], strings.ReplaceAll(parts[1], "\\", "/")
        if !strings.HasPrefix(dest, "/") { dest = "/" + dest }
        data, err := os.ReadFile(host); if err != nil { return err }
        if len(data) == 0 { return fail("empty file is not useful for test injection: %s", host) }
        if uint64(len(data)) > (diskSectors-metaRingSectors-dataStartSector)*sectorSize { return fail("file does not fit: %s", host) }
        if err := ensureDir(c, filepath.ToSlash(filepath.Dir(dest)), now); err != nil { return err }
        idx := findEntry(c, dest)
        if idx < 0 {
            for i := 0; i < maxFiles; i++ { if c[headerSize+i*entrySize] == 0 { idx = i; break } }
        }
        if idx < 0 { return errors.New("storage catalog has no free file entry") }
        e := c[headerSize+idx*entrySize:headerSize+(idx+1)*entrySize]
        for i := range e { e[i] = 0 }
        sectors := (uint64(len(data)) + sectorSize - 1) / sectorSize
        if next+sectors >= diskSectors-metaRingSectors { return errors.New("storage data area is full") }
        e[0], e[1] = 1, 0; put64(e, 4, next); put64(e, 12, sectors); put64(e, 20, uint64(len(data)))
        put64(e, 28, now); put64(e, 36, now); put64(e, 44, now); pathBytes(e[52:180], dest)
        put64(e, entryChecksumOffset, entryChecksum(e))
        padded := make([]byte, int(sectors)*sectorSize); copy(padded, data)
        if err := writeAtFull(f, int64(next*sectorSize), padded); err != nil { return err }
        fmt.Printf("injected %s -> %s (%d bytes, sectors %d)\n", host, dest, len(data), sectors)
        next += sectors; addedBytes += uint64(len(data))
    }
    put64(c, 48, usedEntryCount(c)); put64(c, 56, u64(c, 56)+addedBytes)
    put64(c, checksumOffset, 0); put64(c, checksumOffset, catalogChecksum(c))
    if err := writeAtFull(f, int64(catalogSector*sectorSize), c); err != nil { return err }
    if err := writeAtFull(f, int64(backupCatalogSector*sectorSize), c); err != nil { return err }
    verify, err := loadCatalog(f); if err != nil { return err }
    if u64(verify, 48) != usedEntryCount(verify) { return errors.New("post-write catalog verification failed") }
    fmt.Printf("storage injection verified: %s (%d entries)\n", image, u64(verify, 48))
    return nil
}

func main() {
    image := flag.String("image", "build/storage.img", "C-OS raw storage image")
    var additions addList
    flag.Var(&additions, "add", "host-file=/path/in/C-OS (repeatable)")
    flag.Parse()
    if len(additions) == 0 { fmt.Fprintln(os.Stderr, "usage: inject_storage -image build/storage.img -add sample.mp3=/music/sample.mp3"); os.Exit(2) }
    if err := inject(*image, additions); err != nil { fmt.Fprintln(os.Stderr, "storage injection failed:", err); os.Exit(1) }
}
